Added overview concept guide to all ways to extend

erictune opened the pull request:
Added overview concept guide to all ways to extend kubernetes.

Builds on https://github.com/kubernetes/website/pull/6220

To be followed by more detail on "Automation" and on "User-defined types".

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/website/6224)
<!-- Reviewable:end -->

erictune commented:
@chenopis @lavalamp

ahmetb commented:
I think this should go to `/docs/reference`. Concepts is for everyone, we don’t expect most people to write extensions.

ahmetb commented:
I think this should go to `/docs/reference`. Concepts is for everyone, we don’t expect most people to write extensions.

ahmetb commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,210 @@
+<!-- image source drawing: https://docs.google.com/drawings/d/1sdviU6lDz4BpnzJNHfNpQrqI9F19QZ07KnhnxVrp2yg/edit --> 
+
+## API Extensions
+### User-Defined Types
```
is this CRD? if so might be good to use the same terminology.

erictune replied:
Yes, Fixed.

ahmetb commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,210 @@
+
+Do not use User-defined Types as data storage for application, user, or monitoring data.
+
+For more about User-Defined Types, see **Link Here**.
```
link missing

erictune replied:
fixed.

ahmetb commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,210 @@
+
+### Combining New APIs with Automation
+
+Often, when you add a new API, you also add a control loop that reads and/or writes the new APIs.  This is often called the **Operator** pattern.
```
a common question is "What is an Operator". I don't think we don't have any docs on this. this might be a good place to add it (+ to the glossary perhaps, too).

erictune replied:
I'm working on some, and have made a note to add a link to it here once it is added.

enisoc replied:
According to [CoreOS](https://coreos.com/operators/), a custom resource and a custom controller are not enough for something to count as an Operator. Their definition also includes a third factor:

> An Operator is an application-specific controller that extends the Kubernetes API to create, configure and manage instances of complex stateful applications on behalf of a Kubernetes user. It builds upon the basic Kubernetes resource and controller concepts, but also includes domain or application-specific knowledge to automate common tasks better managed by computers.

So for example, if I write a CRD + custom controller for BlueGreenDeployment, that's not an Operator. It's not app-specific, and it's not about managing complex stateful apps.

erictune replied:
Adjusted the text:
> Often, when you add a new API, you also add a control loop that reads and/or writes the new APIs.  When the combination of a Custom API and a control loop is used to manage a specific, usually stateful, application, this is called the **Operator** pattern.  Custom APIs and control loops can also be used to control other resources, such as storage, policies, and so

ahmetb commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,210 @@
+
+### Authorization
+
+ [Authorization](https://kubernetes.io/docs/admin/authorization/webhook/) determines whether specific users can read, write, and do other operations on API resources.  It just works at the level of whole resources -- it doesn't discriminate based on arbitrary object fields.  If the built-in authorization options don't meet your needs, and [Authorization webhook](https://kubernetes.io/docs/admin/a ...
```
replace all `https://kubernetes.io/` links with `/`

erictune replied:
fixed.

verult commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,210 @@
+*   The apiserver serves various kinds of **resources**, ③.  **Built-in resource kinds**, like `pods`, are defined by the Kubernetes project and have to be .     You can also add resources that you definde are several ways to add **Custom Resources** to your API-Server.  Custom Resources are often used with API Access Extensions.
+*   The kubernetes scheduler, ④, decides which nodes to place pods on.  There are several ways to extend scheduling.  These are described in the [Scheduler Extensions](bloop) section.
+*   Much of the behavior of Kubernetes is implemented by client programs Controllers, ⑤.  You can write your own, as described in the [Controllers](boo) section.  Controllers are often used in conjunction with User-defined Types.
+*   The kubelet runs on servers, and mediates access to storage, ⑥, and the network, ⑦.  These are described in the [Infrastructure Extensions](poof) section.
```
(6) and (7) are flipped

erictune commented:
@ahmetb I first thought that this should go in `docs/reference` too.  

I talked to @chenopis wo told me that concept vs reference it is not about how many people need to know it.  It is about whether the person is still deciding what features to use (concepts) or if they already know what feature to use, and just need to see all of the options/fields of how to use it (reference).

erictune commented:
@ahmetb thanks for review.
@verult thanks for review, fixed swapped numbers.

tpepper commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,209 @@
+
+{% capture overview %}
+
+Kubernetes has many ways to configure and extend it.  As a result,
```
Suggest:

"Kubernetes is highly configurable and extensible.  As a result,"

erictune replied:
Thanks, accepted suggestion.

tpepper commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,209 @@
+
+## Configuration
+
+**Configuration files** and **flags** are documented in the Reference section of the online documentation, under each binary: [kubelet](/docs/admin/kubelet/), [kube-apiserver](/docs/admin/kube-apiserver/),  [kube-controller-manager](/docs/admin/kube-controller-manager/), and [kube-scheduler](/docs/admin/kube-scheduler/).  Flags and configuration files may not always be changable in a hosted Kuber ...
```
Having the binaries as a bullet list will make for easier future adds/removes of items, versus a comma separated list.

erictune replied:
suggestion accepted.

tpepper commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,209 @@
+
+**Configuration files** and **flags** are documented in the Reference section of the online documentation, under each binary: [kubelet](/docs/admin/kubelet/), [kube-apiserver](/docs/admin/kube-apiserver/),  [kube-controller-manager](/docs/admin/kube-controller-manager/), and [kube-scheduler](/docs/admin/kube-scheduler/).  Flags and configuration files may not always be changable in a hosted Kuber ...
+
+**Built-in Policy APIs**, such as [ResourceQuota](/docs/concepts/policy/resource-quotas/), [PodSecurityPolicies](/docs/concepts/policy/pod-security-policy/), [NetworkPolicy](/docs/concepts/services-networking/network-policies/) and Role-based Access Control ([RBAC](/docs/admin/authorization/rbac/)), are built-in Kubernetes APIs.  APIs are typically used with hosted Kubernetes services, and with m ...
```
Same here.  Bullet list will make for easier future diff's and review of modifications to the doc.

erictune replied:
I think I prefer not to use bulleted list for short lists of examples.  They break of the paragraph structure and thus the flow.

tpepper replied:
I agree.  It's a trade off between language flow and future maintenance.

Today with the paragraph a single line of text, any future add/remove/modify of an item in the list results in a git level change of the entire paragraph.  Were it list form you'd get a more simple diff

```
-  old short line
+ new short line
```

which is much more reviewable.

enisoc commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,210 @@
+Automation generally works with any Kubernetes cluster, including hosted
+clusters, and managed installations.
+
+There is a specific patterns for writing client programs that work well with
```
s/patterns/pattern/

erictune replied:
done

enisoc commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,209 @@
+
+*   Kubectl plugins extend the kubectl binary ①.  [Link for 1.8 docs branch: docs/tasks/extend-kubectl/kubectl-plugins ] They only affect the individual user's local environment.  These extension points are described in the [Client Extensions](todo) section.
+*   The apiserver ②  handles all requests.  Several types of extension points in the apiserver allow authenticating requests, or blocking them based on their content, editing content, and handling deletion.  These are described in the [API Access Extensions](todo) section. 
+*   The apiserver serves various kinds of **resources**, ③.  **Built-in resource kinds**, like `pods`, are defined by the Kubernetes project and have to be .     You can also add resources that you definde are several ways to add **Custom Resources** to your API-Server.  Custom Resources are often used with API Access Extensions.
```
It seems like this line got mangled?

erictune replied:
✔️

enisoc commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,209 @@
+*   The apiserver ②  handles all requests.  Several types of extension points in the apiserver allow authenticating requests, or blocking them based on their content, editing content, and handling deletion.  These are described in the [API Access Extensions](todo) section. 
+*   The apiserver serves various kinds of **resources**, ③.  **Built-in resource kinds**, like `pods`, are defined by the Kubernetes project and have to be .     You can also add resources that you definde are several ways to add **Custom Resources** to your API-Server.  Custom Resources are often used with API Access Extensions.
+*   The kubernetes scheduler, ④, decides which nodes to place pods on.  There are several ways to extend scheduling.  These are described in the [Scheduler Extensions](bloop) section.
+*   Much of the behavior of Kubernetes is implemented by client programs Controllers, ⑤.  You can write your own, as described in the [Controllers](boo) section.  Controllers are often used in conjunction with Custom Resources.
```
"programs called Controllers"?

erictune replied:
✔️

enisoc commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,209 @@
+
+
+
+<img source="https://docs.google.com/drawings/d/e/2PACX-1vRWXNNIVWFDqzDY0CsKZJY3AR8sDeFDXItdc5awYxVH8s0OLherMlEPVUpxPIB1CSUu7GPk7B2fEnzM/pub?w=1440&h=1080">
```
s/source/src/

ahmetb replied:
also probably import the image here :)

erictune replied:
✔️

tpepper commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,209 @@
+
+### Changing Built-in Resources
+
+When you extend the Kubernetes API by adding custom resources, the added Kinds always fall into new API Groups.  You cannot replace or change existing API groups.
```
Suggest either:
...the added **Kind** always falls into a new API Group.
with kind definition.  Or same edit but "kind" instead of "**Kind**".

erictune replied:
You can add several kinds to a new api group.  I've used the word resource instead.

erictune replied:
✔️

tpepper commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,209 @@
+
+Each of these steps offers extension points.  
+
+Kubernetes has several built-in authentication methods that it supports.  I can also sit behind an authenticating proxy, and it can send a token from an Authorization header to a remote service for verification (a webhook).  All of these methods are covered in the [Authentication documentation](/docs/admin/authentication/).
```
s/I can also/It can also/

erictune replied:
done

erictune replied:
✔️

enisoc commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,209 @@
+Plugins](https://github.com/kubernetes/community/blob/master/contributors/devel/flexvolume.md)
+and [Network
+Plugins](/docs/concepts/cluster-administration/network-plugins/))
+and by kubectl.  Plugin interfaces still evolving, and may change.
```
s/still/are still/

erictune replied:
✔️

erictune commented:
I've addressed all review comments.

ahmetb commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,209 @@
+* Learn about [Dynamic admission control](/docs/admin/extensible-admission-controller)
+* Learn more about Infrastructure extensions
+  * [Network Plugins](/docs/concepts/cluster-administration/network-plugin)
+  * [Device Plugins](/docs/concepts/cluster-administration/device-plugins.md    )
```
nit: extra space in link

erictune commented:
This is green.  All outstanding comments addressed, I think.  Hoping for LGTMs from tech and docs.

chenopis commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,217 @@
+
+**Built-in Policy APIs**, such as [ResourceQuota](/docs/concepts/policy/resource-quotas/), [PodSecurityPolicies](/docs/concepts/policy/pod-security-policy/), [NetworkPolicy](/docs/concepts/services-networking/network-policies/) and Role-based Access Control ([RBAC](/docs/admin/authorization/rbac/)), are built-in Kubernetes APIs.  APIs are typically used with hosted Kubernetes services, and with m ...
+
+The remainder of this guide is about Extensions.
```
This does not need to be stated. You can remove this sentence.

chenopis commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,217 @@
+
+Kubernetes is designed to be automated by writing client programs.  Any
+program that reads and/or writes to the Kubernetes API can provide useful
+automation.  **Automation** can run on the cluster or off it.  By following
```
"Automation" does not need to be in bold.

chenopis commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,217 @@
+automation.  **Automation** can run on the cluster or off it.  By following
+the guidance in this doc you can write highly available and robust automation.
+Automation generally works with any Kubernetes cluster, including hosted
+clusters, and managed installations.
```
No comma needed: "...clusters` `and managed installations."

chenopis commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,217 @@
+clusters, and managed installations.
+
+There is a specific pattern for writing client programs that work well with
+Kubernetes called the **Controller** pattern.  Controllers typically read an
```
"Controller" should be in italics instead of in bold. When introducing a new term, the convention is to use italics instead of bold.

chenopis commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,217 @@
+object's `.spec`, possibly do things, and then update the object's `.status`.
+
+A controller is a client of Kubernetes.  When Kubernetes is the client and
+calls out to a remote service, it is called a **Webhook.  **The remote service
```
These should be in italics instead of bold, and make sure the formatting delimiters are in the correct locations, e.g. `*Webhook*`.

chenopis commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,217 @@
+Plugins](https://github.com/kubernetes/community/blob/master/contributors/devel/flexvolume.md)
+and [Network
+Plugins](/docs/concepts/cluster-administration/network-plugins/))
+and by kubectl.  Plugin interfaces are still evolving, and may change.
```
We shouldn't make any forward looking statements such as "Plugin interfaces are still evolving, and may change."

chenopis commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,217 @@
+failure.  
+
+In the webhook model, Kubernetes makes a network request to a remote service.
+In the **Binary Plugin** model, kubernetes executes a binary (program).
```
Italics instead of bold.

chenopis commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,217 @@
+
+<!-- image source diagrams: https://docs.google.com/drawings/d/1k2YdJgNTtNfW7_A8moIIkij-DmVgEhNrn3y2OODwqQQ/view -->
+
+*   Kubectl plugins extend the kubectl binary ①.  [Link for 1.8 docs branch: docs/tasks/extend-kubectl/kubectl-plugins ] They only affect the individual user's local environment.  These extension points are described in the [Client Extensions](todo) section.
```
**Make sure all of the links are no longer "todo" and are correct.**

Also, use an ordered list (numbers instead of bullet points) to match the numbers in the diagram, i.e.:

```
1. Kubectl plugins extend the kubectl binary.  See [Extend kubectl with plugins](/docs/tasks/extend-kubectl/kubectl-plugins/) They only affect the individual user's local environment.  These extension points are described in the [Client Extensions](todo) section.
1. The apiserver handles all requests.  Several types of extension points in the apiserver allow authenticating requests, or blocking them based on their content, editing content, and handling deletion.  These are described in the [API Access Extensions](todo) section.
1. ...
```

Which will render as:

1. Kubectl plugins extend the kubectl binary.  See [Extend kubectl with plugins](/docs/tasks/extend-kubectl/kubectl-plugins/). They only affect the individual user's local environment.  These extension points are described in the [Client Extensions](todo) section.
1. The apiserver handles all requests.  Several types of extension points in the apiserver allow authenticating requests, or blocking them based on their content, editing content, and handling deletion.  These are described in the [API Access Extensions](todo) section.
1. ...

zacharysarah replied:
To @chenopis's excellent review I would add:
- Include only one space after periods.
- Glossary additions are a great idea!

chenopis commented:
Is this `Tech LGTM`?

zacharysarah commented on docs/concepts/overview/extending.md:
```diff
@@ -0,0 +1,217 @@
+project code.  
+
+This guide describes the options for customizing a kubernetes
+cluster.  It is aimed at **Cluster Administrators** who want to
```
Glossary entries are my suggestion as well. ☝️

erictune commented:
I've addressed all comments.  Thank you for them.
I previously had a review of this from @lavalamp when it was in Google Docs form, and I'm fairly confident about the correctness, so I went ahead and gave myself tech LGTM.

erictune commented:
@chenopis I see this live on kubernetes.io which is great.  The glossary tooltips aren't working.  Do those take time or might I have done the glossary stuff wrong?

chenopis commented:
@erictune I'll open a patch and fix the tooltips.
